File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/valid-expect-in-promise.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('valid-expect-in-promise', {} as never, {
  valid: [
    {
      code: `test('case', async () => {
        await load().then(value => expect(value).toBe(1));
      });`,
    },
    {
      code: `test('case', async () => {
        let pending;
        pending ||= load().then(value => expect(value).toBe(1));
        await pending;
      });`,
    },
    {
      code: `test('case', async () => {
        let pending = fallback || load().then(value => expect(value).toBe(1));
        await pending;
      });`,
    },
    {
      code: `test('case', async () => {
        let pending = condition ? load().then(value => expect(value).toBe(1)) : other;
        await pending;
      });`,
    },
    {
      code: `test('case', async () => {
        const pending = [load().then(value => expect(value).toBe(1))];
        await Promise.all(pending);
      });`,
    },
    {
      code: `test('case', async () => {
        const pending = [load().then(value => expect(value).toBe(1))];
        for await (const settled of pending) {
        }
      });`,
    },
    {
      code: `test('case', async () => {
        let pending = load().then(value => expect(value).toBe(1));
        pending ??= fallback;
        await pending;
      });`,
    },
    {
      code: `test('case', () =>
        load().then(value => expect(value).toBe(1))
      );`,
    },
    {
      code: `test('case', () => {
        try {
          return load().then(value => expect(value).toBe(1));
        } catch (error) {}
      });`,
    },
    {
      code: `test('case', async () => {
        try {
          return load().then(value => expect(value).toBe(1));
        } catch (error) {}
      });`,
    },
    {
      code: `test('case', async () => {
        const pending = load().then(value => expect(value).toBe(1));
        unrelated = 1;
        await pending;
      });`,
    },
    {
      code: `test('case', () => {
        const pending = load().then(value => expect(value).toBe(1));
        return Promise.all([pending]);
      });`,
    },
    {
      code: `test('case', () => {
        const [pending] = [load().then(value => expect(value).toBe(1))];
        return pending;
      });`,
    },
    {
      code: `test('case', () => {
        const pending = load().then(value => expect(value).toBe(1));
        expect(pending).resolves.toBeUndefined();
      });`,
    },
    {
      code: `test('done', done => {
        load().then(value => {
          expect(value).toBe(1);
          done();
        });
      });`,
    },
    {
      code: `function callback() {
        const pending = load().then(value => expect(value).toBe(1));
        return pending;
      }
      test('case', callback);`,
    },
    {
      code: `function handler() {
        queue.then(handler);
      }
      test('case', () => promise.then(handler));`,
    },
    {
      code: `function handler(value) {
        expect(value).toBe(1);
        return queue.then(handler);
      }
      test('case', () => promise.then(handler));`,
    },
    {
      code: `test('case', async () => {
        const pending = load().then(value => expect(value).toBe(1));
        try {
          await pending;
        } catch (error) {
          throw error;
        }
      });`,
    },
    {
      code: `test('case', async () => {
        const pending = load().then(value => expect(value).toBe(1));
        if (!ready) throw new Error('no');
        await pending;
      });`,
    },
    {
      code: `test('case', async () => {
        const pending = load().then(value => expect(value).toBe(1));
        try {
          setup();
          await pending;
        } catch (error) {
          throw error;
        }
      });`,
    },
    {
      code: `try {
        test('case', async () => {
          const pending = load().then(value => expect(value).toBe(1));
          await pending;
        });
      } catch (error) {}`,
    },
    {
      code: `try {
        test('case', () => {
          const pending = load().then(value => expect(value).toBe(1));
          return Promise.resolve(pending);
        });
      } catch (error) {}`,
    },
    {
      code: `try {
        test('case', () => {
          const pending = load().then(value => expect(value).toBe(1));
          return Promise.all([pending]);
        });
      } catch (error) {}`,
    },
    {
      code: `Promise.resolve().then(() => expect(1).toBe(2));`,
    },
  ],
  invalid: [
    {
      code: `test('case', () => {
        list.push(load().then(value => expect(value).toBe(1)));
      });`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
    {
      code: `test('case', async () => {
        const pending = [load().then(value => expect(value).toBe(1))];
        await pending;
      });`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
    {
      code: `test('case', async () => {
        let pending = load().then(value => expect(value).toBe(1));
        pending = pending && fallback;
        await pending;
      });`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
    {
      code: `function leaf(value) {
        expect(value).toBe(1);
      }
      function shared() {
        return queue.then(leaf);
      }
      test('one', () => {
        promise.then(shared);
      });
      test('two', () => {
        promise.then(shared);
      });`,
      errors: [
        { messageId: 'expectInFloatingPromise' },
        { messageId: 'expectInFloatingPromise' },
      ],
    },
    {
      code: `test('case', () => {
        load().then(value => expect(value).toBe(1));
      });`,
      errors: [{ messageId: 'expectInFloatingPromise', line: 2, column: 9 }],
    },
    {
      code: `test('case', () => {
        const pending = load().then(value => expect(value).toBe(1));
        return Promise.reject(pending);
      });`,
      errors: [{ messageId: 'expectInFloatingPromise', line: 2, column: 15 }],
    },
    {
      code: `test('case', () => {
        const pending = load().then(value => expect(value).toBe(1));
        return Promise.allSettled([pending]);
      });`,
      errors: [{ messageId: 'expectInFloatingPromise', line: 2, column: 15 }],
    },
    {
      code: `test('case', async () => {
        const pending = load().then(value => expect(value).toBe(1));
        if (condition) await pending;
      });`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
    {
      code: `test('case', async () => {
        const pending = load().then(value => expect(value).toBe(1));
        try {
          await pending;
        } catch {}
      });`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
    {
      code: `test('case', async () => {
        const pending = load().then(value => expect(value).toBe(1));
        try {
          throw new Error('caught');
        } catch {}
      });`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
    {
      code: `test('case', async () => {
        const pending = load().then(value => expect(value).toBe(1));
        try {
          throw new Error('suppressed');
        } finally {
          return;
        }
      });`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
    {
      code: `function callback() {
        load().then(value => expect(value).toBe(1));
      }
      test('case', callback);`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
    {
      code: `test('case', async () => {
        await first().then(() => {
          second().then(value => expect(value).toBe(1));
        });
      });`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
    {
      code: `test('case', () => {
        load()
          .then(value => expect(value).toBe(1))
          .then(value => expect(value).toBe(2));
      });`,
      errors: 1,
    },
    {
      code: `test('case', () => {
        const [pending] = load().then(value => expect(value).toBe(1));
      });`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
    {
      code: `test('case', () => {
        const [pending] = [load().then(value => expect(value).toBe(1))];
      });`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
    {
      code: `test('case', () => {
        const { pending } = {
          pending: load().then(value => expect(value).toBe(1)),
        };
      });`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
    {
      code: `test('case', () => {
        consume({
          pending: load().then(value => expect(value).toBe(1)),
        });
      });`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
    {
      code: `function handler(value) {
        expect(value).toBe(1);
      }
      test('one', () => { first().then(handler); });
      test('two', () => { second().then(handler); });`,
      errors: 2,
    },
    {
      code: `test('case', async () => {
        let pending;
        pending ||= load().then(value => expect(value).toBe(1));
      });`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
    {
      code: `test('case', async () => {
        let pending = load().then(value => expect(value).toBe(1));
        pending &&= fallback;
        await pending;
      });`,
      errors: [{ messageId: 'expectInFloatingPromise' }],
    },
  ],
});

File: internal/plugins/jest/rules/valid_expect_in_promise/valid_expect_in_promise_test.go
package valid_expect_in_promise_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/valid_expect_in_promise"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestValidExpectInPromiseRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&valid_expect_in_promise.ValidExpectInPromiseRule,
		[]rule_tester.ValidTestCase{
			{Code: `test("case", async () => { await promise.then(value => expect(value).toBe(1)); });`},
			{Code: `test("case", () => promise.then(value => expect(value).toBe(1)));`},
			{Code: `test("case", () => { try { return promise.then(value => expect(value).toBe(1)); } catch (error) {} });`},
			{Code: `test("case", async () => { try { return promise.then(value => expect(value).toBe(1)); } catch (error) {} });`},
			{Code: `test("case", async () => { const pending = promise.then(value => expect(value).toBe(1)); other = 1; await pending; });`},
			{Code: `test("case", () => { const pending = promise.then(value => expect(value).toBe(1)); return Promise.resolve(pending); });`},
			{Code: `test("case", () => { const pending = promise.then(value => expect(value).toBe(1)); return Promise.all([pending]); });`},
			{Code: `test("case", () => { const [pending] = [promise.then(value => expect(value).toBe(1))]; return pending; });`},
			{Code: `test("case", () => { const { pending } = { pending: promise.then(value => expect(value).toBe(1)) }; return pending; });`},
			{Code: `test("case", () => { let pending; [pending] = [promise.then(value => expect(value).toBe(1))]; return pending; });`},
			{Code: `test("case", () => { let pending; ({ pending } = { pending: promise.then(value => expect(value).toBe(1)) }); return pending; });`},
			// A logical assignment stores the chain, only conditionally, so the
			// binding is real and the value it may leave in place survives.
			{Code: `test("case", async () => { let pending; pending ||= promise.then(value => expect(value).toBe(1)); await pending; });`},
			{Code: `test("case", async () => { let pending; pending ??= promise.then(value => expect(value).toBe(1)); await pending; });`},
			{Code: `test("case", async () => { let pending = other; pending &&= promise.then(value => expect(value).toBe(1)); await pending; });`},
			{Code: `test("case", async () => { let pending; await (pending ||= promise.then(value => expect(value).toBe(1))); });`},
			{Code: `test("case", async () => { let pending = promise.then(value => expect(value).toBe(1)); pending ||= other; await pending; });`},
			// Every operator that carries the chain into the binding binds it,
			// the same way the value walks up to a direct await.
			{Code: `test("case", async () => { let pending = fallback || promise.then(value => expect(value).toBe(1)); await pending; });`},
			{Code: `test("case", async () => { let pending = fallback ?? promise.then(value => expect(value).toBe(1)); await pending; });`},
			{Code: `test("case", async () => { let pending = ready && promise.then(value => expect(value).toBe(1)); await pending; });`},
			{Code: `test("case", async () => { let pending = condition ? promise.then(value => expect(value).toBe(1)) : other; await pending; });`},
			{Code: `test("case", async () => { let pending = (setup(), promise.then(value => expect(value).toBe(1))); await pending; });`},
			{Code: `test("case", async () => { let pending = other; pending = pending || promise.then(value => expect(value).toBe(1)); await pending; });`},
			// The longhand of a preserving logical assignment keeps the promise
			// the binding already holds, because a promise is truthy.
			{Code: `test("case", async () => { let pending = promise.then(value => expect(value).toBe(1)); pending = pending || other; await pending; });`},
			// An array in a binding is consumed element-wise.
			{Code: `test("case", async () => { const pending = [promise.then(value => expect(value).toBe(1))]; await Promise.all(pending); });`},
			{Code: `test("case", async () => { const pending = [promise.then(value => expect(value).toBe(1))]; for await (const settled of pending) {} });`},
			{Code: `test("case", async () => { for await (const settled of [promise.then(value => expect(value).toBe(1))]) {} });`},
			{Code: `test("case", () => { const pending = promise.then(value => expect(value).toBe(1)); expect(pending).resolves.toBeUndefined(); });`},
			{Code: `test("case", () => expect(promise.then(value => expect(value).toBe(1))).resolves.toBeUndefined());`},
			{Code: `test("case", () => condition ? promise.then(value => expect(value).toBe(1)) : Promise.resolve());`},
			{Code: `test("case", () => condition && promise.then(value => expect(value).toBe(1)));`},
			{Code: `test("case", async () => { const pending = promise.then(value => expect(value).toBe(1)); if (condition) await pending; else return pending; });`},
			{Code: `test("case", async () => { const pending = promise.then(value => expect(value).toBe(1)); try { await pending; } catch (error) { throw error; } });`},
			{Code: `test("case", async () => { const pending = promise.then(value => expect(value).toBe(1)); if (!ready) { throw new Error("no"); } await pending; });`},
			{Code: `test("case", async () => { const pending = promise.then(value => expect(value).toBe(1)); try { setup(); await pending; } catch (error) { throw error; } });`},
			{Code: `test("case", async () => { const pending = promise.then(value => expect(value).toBe(1)); try { throw new Error("no"); } finally { cleanup(); } });`},
			{Code: `test("case", async () => { const pending = promise.then(value => expect(value).toBe(1)); try { await pending; } finally { cleanup(); } });`},
			{Code: `try { test("case", async () => { const pending = promise.then(value => expect(value).toBe(1)); await pending; }); } catch (error) {}`},
			{Code: `try { test("case", () => { const pending = promise.then(value => expect(value).toBe(1)); return Promise.resolve(pending); }); } catch (error) {}`},
			{Code: `try { test("case", () => { const pending = promise.then(value => expect(value).toBe(1)); return Promise.all([pending]); }); } catch (error) {}`},
			{Code: `test("done", done => { promise.then(value => { expect(value).toBe(1); done(); }); });`},
			{Code: `function callback() { const pending = promise.then(value => expect(value).toBe(1)); return pending; } test("case", callback);`},
			{Code: `function handler(value) { expect(value).toBe(1); } test("one", () => promise.then(handler)); test("two", () => promise.then(handler));`},
			{Code: `function handler() { queue.then(handler); } test("case", () => promise.then(handler));`},
			{Code: `function first() { queue.then(second); } function second() { queue.then(first); } test("case", () => promise.then(first));`},
			{Code: `function handler(value) { expect(value).toBe(1); return queue.then(handler); } test("case", () => promise.then(handler));`},
			{Code: `let handler; test("case", () => promise.then(handler));`},
			{Code: `declare const handler: () => void; test("case", () => promise.then(handler));`},
			{Code: `let handler: (value: unknown) => void; beforeEach(() => { handler = value => expect(value).toBe(1); }); test("case", () => promise.then(handler));`},
			{Code: `test("done", done => { test("nested", () => {}); promise.then(value => { expect(value).toBe(1); done(); }); });`},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code: `test("case", () => { promise.then(value => expect(value).toBe(1)); });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
					Line:      1,
					Column:    22,
					EndColumn: 67,
				}},
			},
			{
				Code: `test("case", () => { const pending = promise.then(value => expect(value).toBe(1)); return Promise.reject(pending); });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
					Line:      1,
					Column:    28,
				}},
			},
			{
				Code: `test("case", () => { const pending = promise.then(value => expect(value).toBe(1)); return Promise.allSettled([pending]); });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
					Line:      1,
					Column:    28,
				}},
			},
			{
				Code: `function callback() { promise.then(value => expect(value).toBe(1)); } test("case", callback);`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
					Line:      1,
					Column:    23,
				}},
			},
			{
				Code: `test("case", async () => { await first.then(() => { second.then(value => expect(value).toBe(1)); }); });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
					Line:      1,
					Column:    53,
				}},
			},
			{
				Code: `test("case", async () => { const pending = promise.then(value => expect(value).toBe(1)); if (condition) await pending; });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
					Line:      1,
					Column:    34,
				}},
			},
			{
				Code: `test("case", () => { if (condition) { promise.then(value => expect(value).toBe(1)); } });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
					Line:      1,
					Column:    39,
				}},
			},
			{
				Code: `test("case", () => { const pending = promise.then(value => expect(value).toBe(1)); return Promise.all([[pending]]); });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			{
				Code: `test("case", () => { const [pending] = [promise.then(value => expect(value).toBe(1))]; log(pending); });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			{
				Code: `test("case", () => { let pending; [pending] = [promise.then(value => expect(value).toBe(1))]; log(pending); });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			{
				Code: `test("case", () => { let pending; ({ pending } = { pending: promise.then(value => expect(value).toBe(1)) }); log(pending); });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			{
				Code: `test("case", async () => { let pending; pending ||= promise.then(value => expect(value).toBe(1)); });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
					Line:      1,
					Column:    41,
					EndColumn: 97,
				}},
			},
			// A chain stored by a logical assignment floats from the
			// assignment, and one handed to another call from its statement.
			{
				Code: `test("case", async () => { obj.pending ||= promise.then(value => expect(value).toBe(1)); });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
					Line:      1,
					Column:    28,
					EndColumn: 88,
				}},
			},
			{
				Code: `test("case", () => { list.push(promise.then(value => expect(value).toBe(1))); });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
					Line:      1,
					Column:    22,
					EndColumn: 78,
				}},
			},
			{
				Code: `test("case", () => { helper(promise.then(value => expect(value).toBe(1))); });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
					Line:      1,
					Column:    22,
					EndColumn: 75,
				}},
			},
			// Awaiting an array does not await its elements.
			{
				Code: `test("case", async () => { const pending = [promise.then(value => expect(value).toBe(1))]; await pending; });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			// `p = p && other` always replaces the promise p holds.
			{
				Code: `test("case", async () => { let pending = promise.then(value => expect(value).toBe(1)); pending = pending && other; await pending; });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			// A promise is truthy, so `&&=` always replaces the one p holds.
			{
				Code: `test("case", async () => { let pending = promise.then(value => expect(value).toBe(1)); pending &&= other; await pending; });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			// A compound arithmetic assignment does not store the promise.
			{
				Code: `test("case", async () => { let count = 0; count += promise.then(value => expect(value).toBe(1)); await count; });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			{
				Code: `test("case", () => { promise.then(value => expect(value).toBe(1)).then(value => expect(value).toBe(2)); });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			{
				Code: `function handler(value) { expect(value).toBe(1); } test("one", () => { promise.then(handler); }); test("two", () => { promise.then(handler); });`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "expectInFloatingPromise"},
					{MessageId: "expectInFloatingPromise"},
				},
			},
			{
				Code: `function leaf(value) { expect(value).toBe(1); } function shared() { return queue.then(leaf); } test("one", () => { promise.then(shared); }); test("two", () => { promise.then(shared); });`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "expectInFloatingPromise"},
					{MessageId: "expectInFloatingPromise"},
				},
			},
			{
				Code: `function leaf(value) { expect(value).toBe(1); } function shared() { queue.then(leaf); } test("one", () => promise.then(shared)); test("two", () => promise.then(shared));`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "expectInFloatingPromise"},
				},
			},
			{
				Code: `test("case", async () => { const pending = promise.then(value => expect(value).toBe(1)); while (condition) { await pending; } });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			{
				Code: `test("case", () => { const pending = promise.then(value => expect(value).toBe(1)); return condition ? pending : other; });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			{
				Code: `test("case", () => { const pending = promise.then(value => expect(value).toBe(1)); return condition && pending; });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			{
				Code: `test("case", async () => { const pending = promise.then(value => expect(value).toBe(1)); try { await pending; } catch {} });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			{
				Code: `test("case", async () => { const pending = promise.then(value => expect(value).toBe(1)); try { await pending; } finally { return; } });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			{
				Code: `test("case", async () => { const pending = promise.then(value => expect(value).toBe(1)); try { throw new Error("caught"); } catch {} });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
			{
				Code: `test("case", async () => { const pending = promise.then(value => expect(value).toBe(1)); try { throw new Error("suppressed"); } finally { return; } });`,
				Errors: []rule_tester.InvalidTestCaseError{{
					MessageId: "expectInFloatingPromise",
				}},
			},
		},
	)
}
